File: frontend/src/lib/company.ts
import { apiClient } from "./api";

export type CompanyUserRole = "ADMIN" | "USER" | "VIEWER";

export interface Company {
  id: string;
  name: string;
  short_code: string;
  registration_no?: string | null;
  tax_pin?: string | null;
  country: string;
  base_currency: string;
  address?: string | null;
  phone?: string | null;
  email?: string | null;
  website?: string | null;
  logo_url?: string | null;
  is_active: boolean;
  is_default: boolean;
  created_at: string;
  branch_count: number;
  user_count: number;
}

export interface Branch {
  id: string;
  company_id: string;
  name: string;
  branch_code: string;
  branch_type?: string | null;
  address?: string | null;
  city?: string | null;
  phone?: string | null;
  is_active: boolean;
  is_default: boolean;
  created_at: string;
}

export interface UserAccess {
  id: string;
  user_id: string;
  company_id: string;
  role: CompanyUserRole;
  is_default: boolean;
  username?: string | null;
  full_name?: string | null;
  created_at: string;
}

export interface CompanySummary {
  company_id: string;
  company_name: string;
  short_code: string;
  base_currency: string;
  branch_count: number;
  user_count: number;
  total_budgeted: number;
  open_po_count: number;
  open_so_count: number;
  product_count: number;
  warehouse_count: number;
}

const BASE = "/api/v1/companies";

export const companyApi = {
  async list(): Promise<Company[]> {
    const r = await apiClient.get<Company[]>(`${BASE}/`);
    return r.data;
  },
  async create(data: {
    name: string; short_code: string; registration_no?: string; tax_pin?: string;
    country?: string; base_currency?: string; address?: string; phone?: string; email?: string;
  }): Promise<Company> {
    const r = await apiClient.post<Company>(`${BASE}/`, data);
    return r.data;
  },
  async get(id: string): Promise<Company> {
    const r = await apiClient.get<Company>(`${BASE}/${id}`);
    return r.data;
  },
  async update(id: string, data: Partial<Company>): Promise<Company> {
    const r = await apiClient.patch<Company>(`${BASE}/${id}`, data);
    return r.data;
  },
  async setDefault(id: string): Promise<Company> {
    const r = await apiClient.post<Company>(`${BASE}/${id}/set-default`);
    return r.data;
  },
  async listBranches(companyId: string): Promise<Branch[]> {
    const r = await apiClient.get<Branch[]>(`${BASE}/${companyId}/branches`);
    return r.data;
  },
  async addBranch(companyId: string, data: {
    name: string; branch_code: string; branch_type?: string; address?: string; city?: string; phone?: string;
  }): Promise<Branch> {
    const r = await apiClient.post<Branch>(`${BASE}/${companyId}/branches`, data);
    return r.data;
  },
  async listUsers(companyId: string): Promise<UserAccess[]> {
    const r = await apiClient.get<UserAccess[]>(`${BASE}/${companyId}/users`);
    return r.data;
  },
  async grantAccess(companyId: string, data: { user_id: string; role: CompanyUserRole; is_default?: boolean }): Promise<UserAccess> {
    const r = await apiClient.post<UserAccess>(`${BASE}/${companyId}/users`, data);
    return r.data;
  },
  async revokeAccess(companyId: string, userId: string): Promise<void> {
    await apiClient.delete(`${BASE}/${companyId}/users/${userId}`);
  },
  async getSummary(companyId: string): Promise<CompanySummary> {
    const r = await apiClient.get<CompanySummary>(`${BASE}/${companyId}/summary`);
    return r.data;
  },
};

// ── Local company context ─────────────────────────────────────────────────────

const STORAGE_KEY = "active_company_id";

export function getActiveCompanyId(): string | null {
  if (typeof window === "undefined") return null;
  return localStorage.getItem(STORAGE_KEY);
}

export function setActiveCompanyId(id: string): void {
  localStorage.setItem(STORAGE_KEY, id);
}

export function clearActiveCompanyId(): void {
  localStorage.removeItem(STORAGE_KEY);
}

File: backend/app/schemas/company.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Optional, List
import uuid

from pydantic import BaseModel, ConfigDict

from app.models.company import CompanyUserRole


class BranchCreate(BaseModel):
    name:        str
    branch_code: str
    branch_type: Optional[str] = None
    address:     Optional[str] = None
    city:        Optional[str] = None
    phone:       Optional[str] = None
    is_default:  bool = False


class BranchRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id:          uuid.UUID
    company_id:  uuid.UUID
    name:        str
    branch_code: str
    branch_type: Optional[str]
    address:     Optional[str]
    city:        Optional[str]
    phone:       Optional[str]
    is_active:   bool
    is_default:  bool
    created_at:  datetime


class CompanyCreate(BaseModel):
    name:            str
    short_code:      str
    registration_no: Optional[str] = None
    tax_pin:         Optional[str] = None
    country:         str = "Kenya"
    base_currency:   str = "KES"
    address:         Optional[str] = None
    phone:           Optional[str] = None
    email:           Optional[str] = None
    website:         Optional[str] = None


class CompanyUpdate(BaseModel):
    name:            Optional[str] = None
    registration_no: Optional[str] = None
    tax_pin:         Optional[str] = None
    address:         Optional[str] = None
    phone:           Optional[str] = None
    email:           Optional[str] = None
    website:         Optional[str] = None
    logo_url:        Optional[str] = None
    is_active:       Optional[bool] = None


class CompanyRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id:              uuid.UUID
    name:            str
    short_code:      str
    registration_no: Optional[str]
    tax_pin:         Optional[str]
    country:         str
    base_currency:   str
    address:         Optional[str]
    phone:           Optional[str]
    email:           Optional[str]
    website:         Optional[str]
    logo_url:        Optional[str]
    is_active:       bool
    is_default:      bool
    created_at:      datetime
    branch_count:    int = 0
    user_count:      int = 0


class GrantUserAccess(BaseModel):
    user_id:    uuid.UUID
    role:       CompanyUserRole = CompanyUserRole.USER
    is_default: bool = False


class UserAccessRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id:         uuid.UUID
    user_id:    uuid.UUID
    company_id: uuid.UUID
    role:       CompanyUserRole
    is_default: bool
    username:   Optional[str] = None
    full_name:  Optional[str] = None
    created_at: datetime


class CompanySummary(BaseModel):
    company_id:       uuid.UUID
    company_name:     str
    short_code:       str
    base_currency:    str
    branch_count:     int
    user_count:       int
    # Financial snapshots (best-effort from available data)
    total_budgeted:   Decimal = Decimal("0")
    open_po_count:    int = 0
    open_so_count:    int = 0
    product_count:    int = 0
    warehouse_count:  int = 0
